### PR TITLE
wrr: add EDF implementation of weighted round robin.

### PR DESCRIPTION
**Pull request by alazarev**
EDF is strictly better than current random based wrr (balancer/internal/wrr/random.go).
It can be used to load balance between balancer groups (balancer/xds/edsbalancer/balancergroup.go) and
within balancer group using weighted round robin balancer implementation (coming in the next pull request for https://github.com/grpc/grpc-go/issues/2827).

### CODE REVIEW

**Comment by easwars** [comment]
I have a pending PR https://github.com/grpc/grpc-go/pull/2950 to refactor the current xds code in grpc/balancer/xds to grpc/xds/balancer.

Do you plan to use the EDF implementation in this PR from outside of xds code? Please let me know.

**Comment by alazarev** [comment]
@easwars Yes, I'm going to use EDF implementation from balancer/weightedroundrobin/weightedroundrobin.go

**Comment by easwars** [comment]
Thanks @alazarev for your reply.
I will modify my PR to move current grpc/balancer/internal/wrr to grpc/internal/wrr so that it can be used both from grpc/balancer/weightedroundrobin and grpc/xds/balancer.

**Comment by menghanl** [comment]
@alazarev #2950 moved the files/directories. Can you update this PR? Thanks!

**Comment by alazarev** [comment]
Picked up #2950 changes.

File: internal/wrr/edf.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	heap.Push(&edf.items, &entry)
+}
+
+func (edf *edfWrr) Next() interface{} {
+	if len(edf.items) == 0 {
+		return nil
+	}
+	item := edf.items[0]
+	item.deadline = edf.currentTime() + 1.0/float64(item.weight)
+	heap.Fix(&edf.items, 0)
```
Comment (menghanl) [inline internal/wrr/edf.go]: Is this thread safe?

File: internal/wrr/edf.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// edfWrr is a struct for EDF weighted round robin implementation.
+type edfWrr struct {
+	items edfPriorityQueue
+}
+
+// NewEdfWrr creates Earliest Deadline First (EDF)
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) implementation for weighted round robin.
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at current time + 1 / weight, providing weighted round robin behavior with O(log n) pick time.
+func NewEdfWrr() WRR {
```
Comment (menghanl) [inline internal/wrr/edf.go]: Nit: `NewEDF()`?

EDF is an acronym. And this function will be invoked as `wrr.NewEDF()`.

File: internal/wrr/edf.go (hunk 1 above)
Comment (alazarev) [reply internal/wrr/edf.go]: Added lock to make it thread safe

**Comment by alazarev** [comment]
It looks like CI failure on go1.11 is in internal/transport and not related to PR

**Pull request merged**